### PR TITLE
Refactorings

### PR DESCRIPTION
**Pull request by jakearchibald**
This is some stuff I did as I was fixing the second-file-load issue.

Feel free to throw it back at me.

### CODE REVIEW

File: src/components/app/index.tsx [hunk 1]
```diff
@@ -35,6 +35,28 @@ interface State {
...
+    }
+  })();
+
+  const blob = new Blob([compressedData], {
+    type: encoderMap[encodeData.type].mimeType,
+  });
+
+  const bitmap = await createImageBitmap(blob);
+  return bitmap;
+}
```
Comment (jakearchibald) [inline src/components/app/index.tsx]: This was only using state to set the error message, so I moved that to `updateImage`.

File: src/components/app/index.tsx [hunk 2]
```diff
@@ -95,7 +117,7 @@ export default class App extends Component<Props, State> {
     const { source, images } = this.state;
 
     for (const [i, image] of images.entries()) {
-      if (source !== prevState.source || image !== prevState.images[i]) {
+      if (source !== prevState.source || image.encoderState !== prevState.images[i].encoderState) {
```
Comment (jakearchibald) [inline src/components/app/index.tsx]: The image only needs updated if the encoder settings have changed (or the source has changed)

File: src/components/app/index.tsx [hunk 3]
```diff
@@ -133,57 +156,53 @@ export default class App extends Component<Props, State> {
...
 
-    image = this.state.images[index];
-    // If a later encode has landed before this one, return.
-    if (loadingCounter < image.loadedCounter) return;
-    image.bmp = result;
-    image.loading = image.loadingCounter !== loadingCounter;
-    image.loadedCounter = loadingCounter;
-    this.setState({ });
-  }
+    this.setState({ images });
```
Comment (jakearchibald) [inline src/components/app/index.tsx]: I'm trying to avoid editing state & instead setting new state.

File: src/components/output/index.tsx [hunk 4]
```diff
@@ -39,13 +39,17 @@ export default class Output extends Component<Props, State> {
     }
   }
 
+  shouldComponentUpdate(nextProps: Props) {
+    return this.props.leftImg !== nextProps.leftImg || this.props.rightImg !== nextProps.rightImg;
+  }
```
Comment (jakearchibald) [inline src/components/output/index.tsx]: I did this to reduce calls to render() while I was debugging. Is it worth keeping around?

File: src/lib/util.ts [hunk 5]
```diff
@@ -44,9 +44,9 @@ export async function bitmapToImageData(bitmap: ImageBitmap): Promise<ImageData>
 }
 
 /** Replace the contents of a canvas with the given bitmap */
-export function drawBitmapToCanvas (canvas: HTMLCanvasElement, img: ImageBitmap) {
+export function drawBitmapToCanvas(canvas: HTMLCanvasElement, bitmap: ImageBitmap) {
   const ctx = canvas.getContext('2d');
   if (!ctx) throw Error('Canvas not initialized');
   ctx.clearRect(0, 0, canvas.width, canvas.height);
-  ctx.drawImage(img, 0, 0);
+  ctx.drawImage(bitmap, 0, 0);
```
Comment (jakearchibald) [inline src/lib/util.ts]: Just making the naming consistent with the previous function in this file.

File: src/components/app/index.tsx [hunk 6]
```diff
@@ -35,6 +35,28 @@ interface State {
   error?: string;
 }
 
+async function updateCompressedImage(
```
Comment (surma) [inline src/components/app/index.tsx]: Can this be called `compressImage`?

File: src/components/app/index.tsx (hunk 2 above)
Comment (surma) [reply src/components/app/index.tsx]: Seems like your comment should be a code comment

File: src/components/app/index.tsx (hunk 3 above)
Comment (developit) [reply src/components/app/index.tsx]: Was going to suggest that in the other PR, glad its added here.

**Pull request merged**